glass clustering flow + util scripts

Juan-Mateos opened the pull request:
Closes #19 

---

Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have run `flake8` and addressed any linter erors
- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have rebased onto `dev` (or merged any new changes from `dev`)
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [x] Appropriate information has been added to `README`s
- [x] I have explained the feature in this PR or (better) in `output/reports/`
- [x] I have requested a code review

Commit:
glass clustering flow + util scripts

Commit:
updated scripts and readmes

Juan-Mateos commented:
**What this does**

We cluster company descriptions in SIC codes. Flow and utilities scripts can be found in `industrial_taxonomy/pipeline/cluster_glass`.

The user can choose the minimum sector size to include in the clustering (defaults to 1000)

The flow includes branching for different values of an `assigned_shares` parameter that controls the number of companies within a SIC to assign into clusters. We can test performance for different parameter values downstream.

**Before running**
* Install `graph_tool`
* Install `topsbm` by running `gh repo clone martingerlach/hSBM_Topicmodel` inside `industrial_taxonomy/pipeline/glass_clusters` (I couldn't import the topic model when installing via pip @bishax). 
  * You will also need to fix a bug in the `model.clusters` method in the `sbmtm.py` script (see [here](https://github.com/martingerlach/hSBM_Topicmodel/issues/42) for more information).

**Running**
* The full run including a default grid search that extracts the top `[10,100,500,0.5,"all"]` business descriptions for cluster took ca. 3.5 hours in my machine. If you just want to check that the code runs for you, run the flow with...
  * `--test-mode=True` so it only runs on three SIC codes, 
  * `--assigned_shares='[10,1000']` so that only does two extractions per SIC code.
  * This should take 5 minutes or so
 * And get the full extracted data from the latest run of the flow.

Commit:
changes in scripts

Commit:
changed assign_cluster parametre

Commit:
updated readme

bishax commented on README.md:
```diff
@@ -9,6 +9,8 @@
   - Configure pre-commit
   - Configure metaflow to use AWS
+- Run `gh repo clone martingerlach/hSBM_Topicmodel` inside `industrial_taxonomy/pipeline/glass_clusters` to clone the `topsbm` repo.
+  - (Provisional): Fix bug in the `model` method as documented [here](https://github.com/martingerlach/hSBM_Topicmodel/tree/develop)
```
A few points here:
- [General to DAP] We need a process for assessing usability and robustness of new libraries before adopting them - research code like this would not pass this bar
- [Reproducibility] Rather than relying on manual setup we should leverage scripts. In this case all that is actually needed is [one file](https://github.com/martingerlach/hSBM_Topicmodel/blob/master/sbmtm.py) so we can always commit this directly. Relying on User's to patch manually cloned repos is even worse.
- Ages ago when `research_daps` was a thing I started [refactoring topsbm](https://github.com/nestauk/research_daps/blob/24_topsbm/research_daps/flows/topsbm/sbmtm.py) into something slightly more readable and fixing some bugs. It **may** be a better idea to use this file instead?

bishax commented on industrial_taxonomy/pipeline/glass_clusters/flow.py:
```diff
@@ -0,0 +1,129 @@
+            for sector, corp in zip(
+                list(sector_corpora.keys()), sector_corpora.values()
+            )
+        ]
```
I think all this is just...

```suggestion
        all_sectors_corpora = list(sector_corpora.items())
```
?

bishax replied:
Is there any reason why it can't remain as a `dict`?

If the answer is for the test_mode truncation then you can do

```python
from toolz import take

self.sectors_corpora = dict(take(3, sectors_corpora.items())
```

bishax commented on industrial_taxonomy/pipeline/glass_clusters/flow.py:
```diff
@@ -0,0 +1,129 @@
+            )
+        ]
+
+        if self.test_mode is True:
```
`--production run --test-mode=True` shouldn't use shorter data, `--production` should always do a full run.

Ref: [kuebiko](https://nestauk.github.io/kuebiko/pycco/kuebiko/pipeline/nspl/flow.html#section-8)

bishax commented on industrial_taxonomy/pipeline/glass_clusters/flow.py:
```diff
@@ -0,0 +1,129 @@
+        help="share of companies to assign",
+        type=JSONType,
+        default='[10,100,500,0.5,"all"]',
+    )
```
I'm not sure what this parameter is

bishax commented on industrial_taxonomy/pipeline/glass_clusters/topic_utils.py:
```diff
@@ -0,0 +1,68 @@
+
+    return fit_model(
+        corpus=list(sector_corpus.values()), doc_ids=list(sector_corpus.keys())
+    )
```
I would merge these into one function

bishax commented on industrial_taxonomy/pipeline/glass_clusters/flow.py:
```diff
@@ -0,0 +1,129 @@
+@project(name="industrial_taxonomy")
+class ClusterGlass(FlowSpec):
+    """Create sectoral corpora of glass descriptions
+    and cluster them into text sectors
```
Some more methodological info would be useful.

bishax commented on industrial_taxonomy/pipeline/glass_clusters/flow.py:
```diff
@@ -0,0 +1,129 @@
+
+    Attributes:
+        min_sector_size: minimum sector size
+        assigned_shares = share of companies to assign to cluster
```
`=` to `:`

bishax commented on industrial_taxonomy/pipeline/glass_clusters/utils.py:
```diff
@@ -0,0 +1,122 @@
+    )
+
+    return pipe(
+        get_description_tokens(),
```
Input data should be a parameter.

bishax replied:
By input data should be a parameter I mean that you pass the glass-SIC lookup and description tokens into the function as parameters rather than combining data-fetching and processing.

bishax commented on industrial_taxonomy/pipeline/glass_clusters/utils.py:
```diff
@@ -0,0 +1,122 @@
+
+    return pipe(
+        get_description_tokens(),
+        partial(strip_nes),
```
`partial` is not needed for one-parameter function

bishax replied:
What's the methodological rationale for stripping named entities?

bishax commented on industrial_taxonomy/pipeline/glass_clusters/utils.py:
```diff
@@ -0,0 +1,122 @@
+    Returns:
+        dict with sectors and tokenised descriptions for their companies
+    """
+    gl_sic4 = gl_sic4_lookup()
```
The naming isn't obvious, making things slightly more verbose helps a lot

```suggestion
    glass_sic4 = glass_sic4_lookup()
```

bishax replied:
Input data should be a parameter

bishax commented on industrial_taxonomy/pipeline/glass_clusters/flow.py:
```diff
@@ -0,0 +1,129 @@
+    @step
+    def start_parallel_clustering(self):
+        """Creates a branch for every assignment share value"""
+
```
Is there any reason for an empty step?

```suggestion
        self.next(self.start_parallel_clustering)

    @step
    def start_parallel_clustering(self):
        """Creates a branch for every assignment share value"""

```

bishax replied:
Is there any reason for an empty step?

```suggestion

```

bishax commented on industrial_taxonomy/pipeline/glass_clusters/topic_utils.py:
```diff
@@ -0,0 +1,68 @@
+    return model
+
+
+def fit_model_sector(sector_corpus: Dict[str, Dict[int, List[str]]]) -> sbmtm:
```
Bad type? Should this be `Dict[int, List[str]]`?

It may also help to use type aliases, e.g.

```python
DocId = int
Tokens = List[str]

def fit_model_sector(sector_corpus: Dict[DocId, Tokens]) -> sbmtm:
  ...
```

bishax commented on industrial_taxonomy/pipeline/glass_clusters/flow.py:
```diff
@@ -0,0 +1,129 @@
+            for sect, mod in zip(sectors, models)
+        ]
+
+        self.outputs = [sectors, models, clusters]
```
Odd data access pattern that would be much easier to follow if `self.sectors_corpora` was a dict.

```suggestion
        self.sectors = list(self.sectors_corpora.keys())
        self.models = [fit_model_sector(sector_corpus) for sector_corpus in self.sectors_corpora.values()]
        self.clusters = [
            extract_clusters(mod, sect, docs_to_assign=self.input)
            for sect, mod in zip(sectors, models)
        ]
```
Furthermore, we don't need to bundle the artifacts into a single list - if we need to retrieve only one of them in a subsequent step or when doing analysis then that means we have to fetch all the data.

bishax commented on industrial_taxonomy/pipeline/glass_clusters/flow.py:
```diff
@@ -0,0 +1,129 @@
+        from itertools import chain
+
+        # Sectors are the same for all assignment strategies
+        self.sectors = set(chain(*[input.outputs[0] for input in inputs]))
```
Taking the advice on `self.outputs` then this becomes easier to understand

```suggestion
        self.sectors = set(chain.from_iterable(input.sectors for input in inputs))
```

Also: using `.from_iterable` and using a generator comprehension rather than a list comprehension does a bit less intermediate work

bishax commented on industrial_taxonomy/pipeline/glass_clusters/flow.py:
```diff
@@ -0,0 +1,129 @@
+
+    @step
+    def end(self):
+        """Save outputs"""
```
Inaccurate.

bishax commented on industrial_taxonomy/pipeline/glass_clusters/flow.py:
```diff
@@ -0,0 +1,129 @@
+        self.clusters = {
+            f"assigned_{str(param)}": list(chain(*[cl for cl in input.outputs[2]]))
+            for param, input in zip(self.assigned_shares, inputs)
+        }
```
This is quite hard to follow, I'd suggest adding the assigned_shares information within the previous step and only chaining together dicts in this step.

bishax commented on industrial_taxonomy/pipeline/glass_clusters/topic_utils.py:
```diff
@@ -0,0 +1,68 @@
+    elif type(docs_to_assign) is float:
+        extract_n = int(len(model.documents) * docs_to_assign)
+    elif type(docs_to_assign) is str:
+        extract_n = len(model.documents)
```
This should ideally be it's own function.

bishax replied:
```suggestion
    elif docs_to_assign == "all":
        extract_n = len(model.documents)
    else:
        raise ValueError("<msg>")
```

bishax commented on requirements.txt:
```diff
@@ -12,3 +12,4 @@ xlrd
 openpyxl
 metaflow_extensions@git+https://github.com/nestauk/metaflow_extensions@25_fix_project_env
 pyyaml
+git+https://github.com/TopSBM/topsbm
```
Does this package look promising?

bishax commented on industrial_taxonomy/pipeline/glass_clusters/utils.py:
```diff
@@ -0,0 +1,122 @@
+
+
+def gl_sic4_lookup() -> dict:
+    """Creates a lookup between glass_ids and 4 digit SIC codes"""
```
Docstring should indicate that it's a lookup of the first ranked SIC code

bishax commented on industrial_taxonomy/pipeline/glass_clusters/utils.py:
```diff
@@ -0,0 +1,122 @@
+    return {v: k for k, v in _dict.items()}
+
+
+def gl_sic4_lookup() -> dict:
```
Would it be worth putting this as a getter inside `glass_house.py`? It's likely other analyses/pipelines will need the lookup?

bishax commented on industrial_taxonomy/pipeline/glass_clusters/utils.py:
```diff
@@ -0,0 +1,122 @@
+            glass_id=lambda df: df.index.map(
+                reverse_dict(glass_companies_house_lookup())
+            ),
+            sic4=lambda df: df["SIC5_code"].astype(str).apply(lambda x: x[:-1]),
```
This surfaces a bug in the `get_sector` getter! `SIC5_code` is set as type `int` when it should be `str`!!!

bishax replied:
You can use https://pandas.pydata.org/docs/reference/api/pandas.Series.str.slice.html which is faster than `.apply(...)`

bishax replied:
I'd suggest `sic4` -> `SIC4_code` to keep naming conventions consistent.

bishax commented on industrial_taxonomy/pipeline/glass_clusters/utils.py:
```diff
@@ -0,0 +1,122 @@
+            sic4=lambda df: df["SIC5_code"].astype(str).apply(lambda x: x[:-1]),
+        )
+        .query("rank == 1")
+        .drop(axis=1, labels=["SIC5_code", "rank"])
```
I think this line is redundant because you access only `["sic4"]` a few lines down

bishax commented on industrial_taxonomy/pipeline/glass_clusters/utils.py:
```diff
@@ -0,0 +1,122 @@
+    return (
+        get_sector()
+        .assign(
+            glass_id=lambda df: df.index.map(
```
I'd suggest `glass_id` -> `org_id` to keep naming conventions consistent (even if `org_id` is probably a worse name)

bishax commented on industrial_taxonomy/pipeline/glass_clusters/utils.py:
```diff
@@ -0,0 +1,122 @@
+        filtered tokenised descriptions dict
+    """
+
+    return {_id: tok for _id, tok in tokenised.items() if _id in matched_ids}
```
```suggestion
    return {id_: tok for id_, tok in tokenised.items() if id_ in matched_ids}
```
Prefixing with `_` is a signal that a variable is for internal use. I'm guessing you're just trying to avoid shadowing the built-in `id`?

bishax replied:
Applies to `_dict` in this file too

bishax commented on industrial_taxonomy/pipeline/glass_clusters/utils.py:
```diff
@@ -0,0 +1,122 @@
+    """
+    gl_sic4 = gl_sic4_lookup()
+    selected_sectors = set(
+        [
+            sector
+            for sector, sector_n in pd.Series(gl_sic4).value_counts().items()
+            if sector_n > min_sector_size
+        ]
```
```suggestion
            sector
            for sector, sector_n in pd.Series(gl_sic4).value_counts().items()
            if sector_n > min_sector_size
```

bishax commented on industrial_taxonomy/pipeline/glass_clusters/utils.py:
```diff
@@ -0,0 +1,122 @@
+    return {_id: tok for _id, tok in tokenised.items() if _id in matched_ids}
+
+
+def sector_tokens_lookup(
```
Suggest decomposing the lookup creation and the filtering into two functions - c.f. [Single Responsibility Principle](https://en.wikipedia.org/wiki/Single-responsibility_principle)

bishax replied:
If not decomposing in situations like this then it's important to be careful about function naming - here the function name doesn't imply it does anything about filtering.
I'd suggest at least prefixing the function name with `big_` or something to try and signal  that other things are happening.

bishax commented on environment.yaml:
```diff
@@ -5,5 +5,6 @@ dependencies:
   - pip
   - python=3.8
   - openblas
+  - graph-tool
```
Why not use Metaflow's conda decorator?

Commit:
address PR review comments

Commit:
add topsbm script

Commit:
changed doc clustering flow description

Commit:
fixed type error in get_sector, moved glass_sic lookup to glass_house getters, put reverse_dict in new utils script

Commit:
removed pip clone of topsbm from requirements

Commit:
final scripts

Juan-Mateos commented:
I did the following:

**Generic stuff**
* Removed the instructions to clone and fix `hsbmtm` from the `README`
* added the `sbmtm.py` repo

`flow.py`
* Added more information about the purpose of the flow
* Fixed typos in docstrings
* Removed empty and unnecessary steps (leftover from a previous version of the flow)
* I separated topic modelling and description clusterings into two steps because it is unnecessary to train a new topic model for each assignment strategy (we only use the strategy to decide how many documents to cluster *inside* a topic model)
* `test-mode` only runs when not with `--production`.
* Kept `all_sectors_corpora` as a dict (converting it to a list was a leftover from an older version of the flow where I branched the flow over sectors)
* This simplified subsequent steps as you suggested
* I clarified the meaning of the `assigned-shares` parameter
* I created `{assigned_shares:models, clusters}` dicts in the `cluster_glass_descriptions` step as suggested and used the `join` step to combine them into a single  dict.

`topic_utils.py`
* Subsumed `fit_model` code within `fit_model_sector`
* Fixed type typos

`utils.py`
* Removed unnecessary `partial` in the tokenising pipe
* Changed `gl_sic4` name
* Moved `glass_sic4_lookup` to the `glass_house` getters and added new info to the docstring
  * Changed variable names and `apply` to `slice` 
  * I also fixed the typing bug in `get_sector`
* Changed variable names in the dict comprehensions

`requirements.txt`
* I removed the `pip install topsbm` thing.

**I didn't do the following**
* I tried using the `conda` decorator to create a virtual environment with `graph-tool` but this raised a `pip install` error so I decided to follow your advice in Kuebiko: "*stick to using your project’s main environment and only use Metaflow’s conda decorators when you definitely need them."*
 * I tried adding AB's refactored version of `sbmtm.py` but that raised an error I assume is linked to subsequent changes in `graph-tool`:
    * Error message: ` minimize_nested_blockmodel_dl() got an unexpected keyword argument 'deg_corr'`
 * I didn't decompose lookup creation and filtering into two functions because lookup creation would have amounted to a thin wrapper doing `{k:v for k,v in dict.items()}`

**Misc**
* The reason to `strip_nes` is because we wouldn't want ngrams including generic NE names to appear in sector names generated later. Admittedly, this is unlikely to happen + we could also strip them when we name the sectors. I don't feel strongly either way.


**Points for clarification**
  * What do you mean by "input data should be a parameter" (to pipe)?

bishax commented on industrial_taxonomy/pipeline/glass_clusters/flow.py:
```diff
@@ -1,24 +1,27 @@
 from typing import Dict, List, Tuple
-from metaflow import FlowSpec, JSONType, project, step, Parameter
+from metaflow import FlowSpec, JSONType, current, project, step, Parameter
 from industrial_taxonomy.pipeline.glass_clusters.hSBM_Topicmodel.sbmtm import sbmtm
```
```suggestion
from sbmtm import sbmtm
```

- This is using the out-dated cloned repo
- `industrial_taxonomy.pipeline.glass_clusters` isn't a module, there's no `__init__.py`

bishax replied:
This one isn't resolved

bishax commented on industrial_taxonomy/pipeline/glass_clusters/flow.py:
```diff
@@ -50,49 +57,59 @@ def start(self):
     def make_sector_corpora(self):
         """Make the sector corpora"""
+        from toolz import take
         from industrial_taxonomy.pipeline.glass_clusters.utils import (
```
```suggestion
        from utils import (
```

bishax commented on industrial_taxonomy/pipeline/glass_clusters/flow.py:
```diff
@@ -50,49 +57,59 @@ def start(self):
     def cluster_glass_descriptions(self):
-        """Cluster glass descriptions using topsbm"""
+        """Cluster glass descriptions using topsbm
+        for each value of the assigned_shares parametre"""
```
Import below this should be `from topic_utils....`

bishax commented on industrial_taxonomy/pipeline/glass_clusters/flow.py:
```diff
@@ -50,49 +57,59 @@ def start(self):
-        ]
-
-        self.outputs = [sectors, models, clusters]
+        self.sectors = list(self.sectors_corpora.keys())
```
This also happens in the step above!

bishax replied:
Duplicate is removed, but this should be in the previous step as it uses no new information / serves no new purpose.

bishax commented on industrial_taxonomy/pipeline/glass_clusters/flow.py:
```diff
@@ -50,49 +57,59 @@ def start(self):
             self.sectors_corpora = all_sectors_corpora
 
-        self.next(self.start_parallel_clustering)
+        self.next(self.fit_topic_models)
```
`fit_topic_models` could be parallelised using foreach so that each sector model is trained in parallel? (then each model can branch again to do the different clusterings)

bishax commented on industrial_taxonomy/pipeline/glass_clusters/flow.py:
```diff
@@ -1,24 +1,27 @@
+                This is main output we will be using downstream
     """
 
     min_sector_size: int
```
`sectors_corpora` is a `dict` not a `list` !

bishax commented on industrial_taxonomy/pipeline/glass_clusters/utils.py:
```diff
@@ -91,7 +66,7 @@ def sector_tokens_lookup(
     """
 
     return {
-        sector: {_id: tok for _id, tok in tokenised.items() if gl_sic4[_id] == sector}
```
> glass_clusters/utils.py:69: error: Value expression in dictionary comprehension has incompatible type "Dict[int, List[str]]"; expected type "List[str]"

bishax commented on industrial_taxonomy/pipeline/glass_clusters/topic_utils.py:
```diff
@@ -55,13 +65,7 @@ def extract_clusters(
-        extract_n = len(model.documents)
+    extract_n = get_n_docs_to_assign(docs_to_assign, len(model.documents))
 
     cluster_assignment = model.clusters(l=cl_level, n=extract_n)
```
below... 

> glass_clusters/topic_utils.py:72: error: List comprehension has incompatible type List[Tuple[str, Any]]; expected List[Tuple[int, str]]

Commit:
added merge artifacts to store models

bishax commented on industrial_taxonomy/utils/utils.py:
```diff
@@ -0,0 +1,8 @@
+"""Shared utils"""
```
Call this something more informative such as `collections.py`

bishax reviewed (commented):
> I tried using the conda decorator to create a virtual environment with graph-tool but this raised a pip install error

What error did you get? This could be a bug.
I managed to get it working (On Linux, I haven't tried on Mac).

>   I tried adding AB's refactored version of sbmtm.py but that raised an error I assume is linked to subsequent changes in graph-tool:
        Error message: minimize_nested_blockmodel_dl() got an unexpected keyword argument 'deg_corr'

I don't know what's going on with that parameter, maybe the graph-tool API changed :shrug: but that parameter doesn't exist in `graph-tool` for that method.


>  The reason to strip_nes is because we wouldn't want ngrams including generic NE names to appear in sector names generated later. Admittedly, this is unlikely to happen + we could also strip them when we name the sectors. I don't feel strongly either way.

So we don't end up using named entities at all? 
It seems worth investigating the effects on results of using them vs. dropping them.

Commit:
addressing PR review changes

Juan-Mateos commented:
**What I have done:**

`flow.py`
* Removed hSBM import
* Relative paths for `util` and `topic_util` imports
* Removed redundant `self.sectors`
* Fixed various typing problems

`utils.py`
* tokenised filtering pipeline takes data instead of a getter
* added `big_` to `sector_tokens_lookup`

`misc`
* Renamed `utils/utils.py` as `collections.py`

**What I haven't done**
* Branching for each sector

bishax commented on industrial_taxonomy/pipeline/glass_clusters/flow.py:
```diff
@@ -0,0 +1,139 @@
+    assigned_shares: List[float]
+    sectors_corpora: Dict[str, List[str]]
+    sectors: List[str]
+    clusters: List[Tuple[int, str]]
```
Should this be `Dict[str, List[Tuple[int, str]]]`? The description in the docstring is also out of date.

bishax replied:
In the output both elements of the tuple are actually strings!

bishax replied:
By the end of the flow it changes to `Dict[str, Dict[str, tuple]]`?

bishax commented on industrial_taxonomy/pipeline/glass_clusters/utils.py:
```diff
@@ -0,0 +1,103 @@
+        if sector_n > min_sector_size
+    )
+
+    tokenised_descriptions = get_description_tokens()
```
> By input data should be a parameter I mean that you pass the glass-SIC lookup and description tokens into the function as parameters rather than combining data-fetching and processing.


By this I meant that the function `make_sector_corpora` should take in the data as arguments.

bishax replied:
i.e. the general pattern should be something like

```python

def processing_fn(data1, ..., datan, param1, ..., paramn):
    # transformation steps (each typically a function)
    return pipe(data1, ..., filter(lambda xs: xs in data2....))
```

Calling getters at the top-level (i.e. flow-level for us) and passing data into the processing function.
A few of the benefits:
- Easier to test the processing function because we can pass in mock/smaller data
- Easier to re-use the processing function because we can chose the data input, e.g. we may want to use `glass_sic4_lookup` or `get_description_tokens` passing a specific flow run into those getters that tests some new functionality such as an improved tokenisation pipeline
- Separation of concerns
- Purity: If the data is parameterised as an input then we know that if we pass the same inputs in then the processing function will ouput the same result; however if the fetching is done in the function then the same input may yield different outputs.

bishax commented on industrial_taxonomy/pipeline/glass_clusters/flow.py:
```diff
@@ -0,0 +1,139 @@
+
+        self.merge_artifacts(inputs, include=["models"])
+
+        self.sectors = set(chain.from_iterable(input.sectors for input in inputs))
```
This is changing the type from the above step.
Can the line `self.sectors = list(self.sectors_corpora.keys())` not just go in the `make_sector_corpora` step (the point at which it is known), and not be modified again?

bishax commented on industrial_taxonomy/pipeline/glass_clusters/flow.py:
```diff
@@ -0,0 +1,139 @@
+        for each value of the assigned_shares parametre"""
+
+        from topic_utils import (
+            extract_clusters,
```
Pointer on `black`: If you leave trailing `,` at the end of lines like this (e.g. after reducing a long list of imports to just one ) then it won't reformat it to be on one short line

bishax commented on industrial_taxonomy/pipeline/glass_clusters/topic_utils.py:
```diff
@@ -0,0 +1,72 @@
+# Utilities to fit topics models and cluster documents
+
+from typing import List, Tuple, Dict, Union
+from industrial_taxonomy.pipeline.glass_clusters.sbmtm import sbmtm
```
Just `from sbmtm import sbmtm`

Commit:
latest round of reviews

Commit:
fixed type for one of the parameters

Juan-Mateos commented:
**What I have done:**

`flow.py`
* Tested with `mypy` and fixed all typing errors
* Updated the docstrings
* Added `glass_sic4_lookup` as a parameter into `make_sector_corpora`
* Removed duplicated `self.sectors` assignment
* Removed comma from `topic_utils` import

`utils.py`
* Turned `glass_sic4_lookup` into a parameter for `make_sector_corpora`

Given issues loading saved models I opted to leave the `sbmtm` imports as they were previously.

I ran the flow on `test-mode` and it works.

Commit:
fixed typing error in assigned_shares

bishax commented on industrial_taxonomy/pipeline/glass_clusters/utils.py:
```diff
@@ -75,16 +72,18 @@ def big_sector_tokens_lookup(
     Returns:
         dict with sectors and tokenised descriptions for their companies
     """
-    glass_sic4 = glass_sic4_lookup()
```
Line 91: `tokenised_descriptions = get_description_tokens()` this is also data that should be a parameter of `make_sector_corpora`

bishax reviewed (approved):
One last thing then GTM

Commit:
make tokenised descriptions an argument in sector corpora

Juan-Mateos commented:
I turned `get_descriptions_tokenised` into an argument into `make_sector_corpora`

I ran it in `test-mode` and it works.

Commit:
added text_sectors script (#32)

* added text_sectors script

* addressed pr review comments
